_episodes/03-pytest.md

```diff
@@ -0,0 +1,193 @@
+  - "I am a Fortran developer, do I need to care about pytest?"
+  - "How can we use Git hooks to make sure we do not commit broken code?"
+objectives:
+  - "Get comfortable with pytest."
+  - "Learn to write a Git hook to automatize testing."
+keypoints:
+  - "pytest collects and runs all test functions starting with `test_`"
+  - "Python and C/C++ have better tooling for automated tests and you can use those also for Fortran projects (via `iso_c_binding`)."
+  - "Git hooks provide a convenient way to *locally* automatize tests."
+---
+
+> ## Exercise [pytest](http://doc.pytest.org) 
+> 
+> In this exercise we will make a simple Python function and use
+> [pytest](http://doc.pytest.org) to test it. We will try to
+> keep things simple so that those who do not use Python can follow.
+> 
+> Create a new directory and change into it:
+> 
+> ```shell
+> $ mkdir pytest-example
+> $ cd pytest-example
+> ```
+> 
+> Then create a file called `example.py` and copy-paste the following code into it:
+> 
+> ```python
+> def add(a, b):
+>     return a + b
+> 
+> 
+> def test_add():
+>     assert add(2, 3) == 5
+>     assert add('space', 'ship') == 'spaceship'
+> ```
+> 
+> This code contains one genuine function and a test function.
+> 
+> Let us try to test it with pytest:
+> 
+> ```shell
+> $ pytest -v example.py
+> 
+> ============================================================ test session starts =================================
+> platform linux -- Python 3.7.2, pytest-4.3.1, py-1.8.0, pluggy-0.9.0 -- /home/user/pytest-example/venv/bin/python3
+> cachedir: .pytest_cache
+> rootdir: /home/user/pytest-example, inifile:
+> collected 1 item
+> 
+> example.py::test_add PASSED
+> 
+> ========================================================= 1 passed in 0.01 seconds ===============================
+> ```
+> 
+> Yay! The test passed!
+> 
+> Let us break the test!
+> 
+> Introduce a code change which breaks the code and check
+> whether pytest detects the change:
+> 
+> ```shell
+> $ pytest -v example.py
+> ============================================================ test session starts =================================
+> platform linux -- Python 3.7.2, pytest-4.3.1, py-1.8.0, pluggy-0.9.0 -- /home/user/pytest-example/venv/bin/python3
+> cachedir: .pytest_cache
+> rootdir: /home/user/pytest-example, inifile:
+> collected 1 item
+> 
+> example.py::test_add FAILED
+> 
+> ================================================================= FAILURES =======================================
+> _________________________________________________________________ test_add _______________________________________
+> 
+>     def test_add():
+> >       assert add(2, 3) == 5
+> E       assert -1 == 5
+> E         --1
+> E         +5
+> 
+> example.py:6: AssertionError
+> ========================================================= 1 failed in 0.05 seconds ==============
+> ```
+{: .challenge}
+
+---
+
+## Note to Fortran developers
```

Comment:
I wrote this one but was always wondering whether this side-note is helpful or just distracting and confusing. Maybe we could completely remove it?

Reply:
Leave it for now, as a separation between the exercises. Someone will find it useful.